src/solvers/h48/thread.h

```diff
@@ -0,0 +1,301 @@
+{
+	pthread_mutex_lock(&dest->mutex);
+	for (int i = src->front; i != src->rear; i = (i + 1) % MAX_QUEUE_SIZE)
+	{
+			dest->tasks[i] = src->tasks[i];
+			dest->tasks[i].depth = depth;
+	}
+	dest->front = src->front;
+	dest->rear = src->rear;
+	dest->tasks_count = src->tasks_count;
+	pthread_cond_broadcast(&dest->cond);
+	pthread_mutex_unlock(&dest->mutex);
+}
+
+STATIC void *
+start_thread(void *arg)
+{
+	task_queue_t *queue = (task_queue_t *)arg;
+	while (1) {
+		pthread_mutex_lock(&queue->mutex);
+		while (queue->tasks_count == 0 && !queue->terminate) {
+			pthread_cond_wait(&queue->cond, &queue->mutex);
+		}
+		if (queue->tasks_count == 0 && queue->terminate) {
+			pthread_mutex_unlock(&queue->mutex);
+			break;
+		}
+
+		if (queue->tasks_count > 0) {
+			dfsarg_solveh48_t task = queue->tasks[queue->front];
+			queue->front = (queue->front + 1) % MAX_QUEUE_SIZE;
+			queue->tasks_count--;
+			queue->active++;
+			pthread_mutex_unlock(&queue->mutex);
+			
+			solve_h48_single(&task, queue);
+
+			pthread_mutex_lock(&queue->mutex);
+			queue->active--;
+
+			if(queue->tasks_count == 0 && queue->active == 0)
+				pthread_cond_signal(&queue->active_cond);
+		}
+		pthread_mutex_unlock(&queue->mutex);
+	}
+	return NULL;
+}
+
+STATIC int64_t
+solve_h48_bfs(dfsarg_solveh48_t *arg_zero, task_queue_t *tq)
+{
+	dfsarg_solveh48_t queue[MAX_QUEUE_SIZE];
+	int front = 0, rear = 0;
+	dfsarg_solveh48_t nextarg;
+	int depth = 0;
+	int nodes_at_current_depth = 1;
+	int nodes_at_next_depth = 0;
+	queue[rear++] = *arg_zero;
+
+	dfsarg_solveh48_t task_pool[MAX_QUEUE_SIZE];
+
+	while (front < rear){
+		dfsarg_solveh48_t arg = queue[front++];
+		nodes_at_current_depth--;
+
+		if (*arg.nsols == arg.maxsolutions)
+			return 1;
+
+		if (issolved(arg.cube)){
+			if (arg.nmoves + arg.npremoves != arg.depth)
+				continue;
+			solve_h48_appendsolution(&arg);
+			continue;
+		}
+
+		arg.nissbranch = MM_NORMAL;
+		uint32_t allowed = allowednextmove_h48(arg.moves, arg.nmoves, arg.nissbranch);
+
+		for (uint8_t m = 0; m < 18; m++){
+			if (allowed & (1 << m)){
+				nextarg = arg;
+				nextarg.nmoves = arg.nmoves + 1;
+				nextarg.moves[arg.nmoves] = m;
+				nextarg.cube = move(arg.cube, m);
+				nextarg.inverse = premove(arg.inverse, m);
+
+				if (nextarg.nmoves == BFS_DEPTH){
+					dfsarg_solveh48_t *task = &task_pool[rear % MAX_QUEUE_SIZE];
+					*task = nextarg;
+					submit_task(tq, *task);
+				} else {
+					queue[rear++] = nextarg;
+					nodes_at_next_depth++;
+				}
+			}
+		}
+		if (nodes_at_current_depth == 0){
+			nodes_at_current_depth = nodes_at_next_depth;
+			nodes_at_next_depth = 0;
+			LOG("Found %" PRId64 " solutions, searching at depth %" PRId8 "\n", *nextarg.nsols, depth++);
+		}
+		if (depth == BFS_DEPTH) return 0;
+	}
+	return 1;
+}
+
+STATIC int64_t
+solve_h48_single(dfsarg_solveh48_t *arg, task_queue_t *tq)
+{
+	dfsarg_solveh48_t nextarg;
+	int64_t ret;
+	uint8_t m;
+
+	if (*arg->nsols == arg->maxsolutions)
```

Comment:
I get a data race here (with line 50). The variable nsols should me mutex-guarded or made atomic.